### PR TITLE
8230833: LabeledSkinBase computes wrong height with ContentDisplay.GRAPHIC_ONLY

### PR DESCRIPTION
**Pull request by karthikpandelu**
Ignore text condition was not considered in `computePrefHeight` method while calculating the Label height.

Added `isIgnoreText` condition in `computePrefHeight` method while calculating Label height.

Tests are present for all the ContentDisplay types. Modified tests which were failing because of the new height value getting calculated.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8230833](https://bugs.openjdk.org/browse/JDK-8230833): LabeledSkinBase computes wrong height with ContentDisplay.GRAPHIC_ONLY

### Reviewers
 * [Ajit Ghaisas](https://openjdk.org/census#aghaisas) (@aghaisas - **Reviewer**)
 * [John Hendrikx](https://openjdk.org/census#jhendrikx) (@hjohn - Committer)
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx pull/996/head:pull/996` \
`$ git checkout pull/996`

Update a local copy of the PR: \
`$ git checkout pull/996` \
`$ git pull https://git.openjdk.org/jfx pull/996/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 996`

View PR using the GUI difftool: \
`$ git pr show -t 996`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/996.diff">https://git.openjdk.org/jfx/pull/996.diff</a>

</details>

### CODE REVIEW

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (hjohn) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java:393]: The logic of this code seems to have changed more than just the fixing bug. Why are the `prefHeight` functions now called with `-1`?  Normally these should respect the content bias of the node involved.

Also, I notice that you check `graphic == null`, while the `isIgnoreGraphic` function checks quite a bit more:

    boolean isIgnoreGraphic() {
        return (graphic == null ||
                !graphic.isManaged() ||
                getSkinnable().getContentDisplay() == ContentDisplay.TEXT_ONLY);
    }

Doing all those operations on `graphic` when for example the `ContentDisplay` is `TEXT_ONLY` seems unnecessary.  Looking at the rest of the code, `graphicHeight` is only used in one branch in your if/elseif/elseif/else.

I also wonder if a simple fix like this would have the same result:

```diff
-       final double textHeight = Utils.computeTextHeight(font, cleanText,
+       final double textHeight = isIgnoreText() ? 0.0 : Utils.computeTextHeight(font, cleanText,
                 labeled.isWrapText() ? textWidth : 0,
                 labeled.getLineSpacing(), text.getBoundsType());
```

Comment (karthikpandelu) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java]: > The logic of this code seems to have changed more than just the fixing bug. Why are the `prefHeight` functions now called with `-1`? Normally these should respect the content bias of the node involved.
> 

I used `-1` while calling `prefHeight` as I tried to make `computePrefHeight` in consistent with `computePrefWidth` method. But looking at the documentation in Node.java file, it looks like width should be used. So I made modifications to use width while calling `prefHeight`.


> Also, I notice that you check `graphic == null`, while the `isIgnoreGraphic` function checks quite a bit more:
> 
> ```
> boolean isIgnoreGraphic() {
>     return (graphic == null ||
>             !graphic.isManaged() ||
>             getSkinnable().getContentDisplay() == ContentDisplay.TEXT_ONLY);
> }
> ```
> 
> Doing all those operations on `graphic` when for example the `ContentDisplay` is `TEXT_ONLY` seems unnecessary. Looking at the rest of the code, `graphicHeight` is only used in one branch in your if/elseif/elseif/else.
> 

I carried forward the conditions which were already present. I believe removing the condition like `graphic == null` or `txt == null` from `isIgnoreGraphic()` or `isIgnoreText()` methods will create more combinations of conditions combined with ContentDisplay value during height calculation and we have to consider which all the height components to be considered in these cases. Since `graphic==null` is as good as considering  ContentDisplay.TEXT_ONLY, I feel it is better to keep this check.

I have updated the code to use `graphicHeight` in other cases of if/else branch.

> I also wonder if a simple fix like this would have the same result:
> 
> ```diff
> -       final double textHeight = Utils.computeTextHeight(font, cleanText,
> +       final double textHeight = isIgnoreText() ? 0.0 : Utils.computeTextHeight(font, cleanText,
>                  labeled.isWrapText() ? textWidth : 0,
>                  labeled.getLineSpacing(), text.getBoundsType());
> ```

This fixes only the issue of height calculation when text height needs to be ignored. But the original if/else branch present, considers text gap when not required. So I believe modification to the if/else branch is necessary.

**Comment by karthikpandelu** [comment]
Expected values of tests listed below have been changed in this PR. Justification for the change is given below.

When text is null or empty, it is considered as ignore text (GRAPHIC_ONLY) condition. So only graphic height along with top and bottom inset value should be considered for height calculation. Hence changed tests are:

1. whenTextIsNullAndGraphicIsSetWithTOPContentDisplay_computePrefHeight_ReturnsRightAnswer
2. whenTextIsEmptyAndGraphicIsSetWithTOPContentDisplay_computePrefHeight_ReturnsRightAnswer
3. whenTextIsNullAndGraphicIsSetWithBOTTOMContentDisplay_computePrefHeight_ReturnsRightAnswer
4. whenTextIsEmptyAndGraphicIsSetWithBOTTOMContentDisplay_computePrefHeight_ReturnsRightAnswer
5. whenTextIsNullAndGraphicIsSetWithTOPContentDisplay_computeMaxHeight_ReturnsRightAnswer
6. whenTextIsEmptyAndGraphicIsSetWithTOPContentDisplay_computeMaxHeight_ReturnsRightAnswer
7. whenTextIsNullAndGraphicIsSetWithBOTTOMContentDisplay_computeMaxHeight_ReturnsRightAnswer
8. whenTextIsEmptyAndGraphicIsSetWithBOTTOMContentDisplay_computeMaxHeight_ReturnsRightAnswer

**Review by aghaisas — APPROVED** [review]
The fix looks good to me!

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (karthikpandelu) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java]: @hjohn please take a look and review the changes made.

**Comment by karthikpandelu** [comment]
@andy-goryachev-oracle please take a look at this PR and let me know if you have any comments on this.

**Review by hjohn — APPROVED** [review]

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java:402]: My only comment is possibly reorganize the logic to avoid unnecessary computation.  
For example, there is no need to compute text width (line 329) if isIgnoreText() is true; 
Similarly, no need to compute graphicWidth on line 333 if isIgnoreGraphic() is true.

(basically, same comment as @hjohn has made earlier)

What do you think, @karthikpandelu ?

Comment (karthikpandelu) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java:402]: Not sure if I didn't understand your comments properly but I'm guessing you are referring to `textHeight` and `graphicHeight` calculation in line 375 and 380 because the line numbers you mentioned are in `computePrefWidth` method and changes are made to height calculation in `computePrefHeight` only.
Since `textHeight` and `graphicHeight` each are used in 3 conditions from line no 386 to 394 I have kept the calculation in the beginning itself. 

On the other hand `textWidth` declared in line 368 might change in line 371, so I can't use `width` directly instead of defining `textWidth` as `width` need to be used while calling other graphic related methods.

Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java:402]: Both `computePrefHeight()` and `computePrefWidth` can be restructured to avoid unnecessary computation (especially since these are very popular objects).

My point is that, for example, textWidth on line 375 (used to compute textHeight:375) is not used if isIgnoreText() == true.  

Similarly, graphicHeight:380 is not used if isIgnoreGraphic() == true, line 386.

Same optimization can be applied to computePrefWidth():315

Just a suggestion, really.

Comment (karthikpandelu) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java:400]: Understood.
I have optimized the code. Please take a look.
Did following optimizations in the code:

- Optimized `if else` conditions by moving the `graphicHeight` and `graphicWidth` value calculation inside the else statement in `computePrefHeight` and `computePrefWidth` methods respectively.
- Removed unused `graphic` variable.
- In `computePrefHeight` method, since a `if` condition was already present which was same as the one used for `padding` calculation, moved `padding` calculation code to the same location.
- In `computePrefWidth` method, replaced multiple return statements with single statement for better readability and did minor optimizations.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (hjohn) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java:331]: minor: missed a space here after `if`

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (hjohn) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java:393]: minor: missing space before `{`

**Review by hjohn — APPROVED** [review]

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (karthikpandelu) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java]: Updated code.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (karthikpandelu) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java]: Updated code.

**Review by hjohn — APPROVED** [review]

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java:328]: textWidth is being computed even if isIgnoreText() is true.  I wonder if we could re-structure the if-else statements below to avoid unnecessary computations?

There is another suggestion - isIgnoreText() is not a simple getter (does a bit of computation).  We probably should use a boolean early in this method to avoid calling it repeatedly.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java:382]: same comment here - textHeight is computed regardless of whether its needed (isIgnoreText() == true),

and use a boolean to avoid multiple calls to isIgnoreText() in the same method.

isIgnoreGraphic, on the other hand, is very light weight, so it's up to you whether to use a local boolean variable in this and the other method or not.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java:365]: cleanText code block is unnecessary  when isIgnoreText() == true

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (karthikpandelu) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java:329]: Previously both `textWidth` and `graphicWidth` were calculated regardless of the condition. In the updated code calculation of `graphicWidth` is moved to `else` block so that one unnecessary calculation is avoided. Since `textWidth` is used in 2 more conditions in `else` block, calculated it before the start of `if` block.

Since `isIgnoreText()` does more than just checking for `ContentDisplay` value, further reordering `if-else` statements generates different result than expected.
So this further leads to question previously raised by @hjohn, if `isIgnoreText()` and `isIgnoreGraphic()` should only check for ContetDisplay type or check for conditions like null/empty graphic or text as well. If this needs to be separated, again it gives raise to more conditions in the `computePrefWidth()` method.

Added boolean in the beginning as suggested.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (karthikpandelu) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java:385]: In the case of `computePrefHeight` as well calculation of `graphicHeight` is moved to `else` block to avoid one unnecessary calculation.
In this method, by checking for ignore text condition before checking for ignore graphic, we can move `cleanText`, `textWidth` and `textHeight` calculation to `else` block instead of only `graphicHeight`. But this again leads to unexpected result because of the checks present in `isIgnoreText()` method. Hence did not re-order `if-else` statements now.

Added boolean for both `isIgnoreGraphic` and `isIgnoreText`.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (karthikpandelu) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java:368]: Since `clenText` is used in `textHeight` calculation, this cannot be avoided before the `if-else` block as `if-else` block is not re-ordered now. Please refer to the previous comment for the details on `computePrefHeight()` method.

**Comment by andy-goryachev-oracle** [comment]
For your convenience, I've added a Label page to the monkey tester
https://github.com/andy-goryachev-oracle/Test/blob/main/src/goryachev/monkey/MonkeyTesterApp.java

(sorry, it is still in a separate repository.  if you use Eclipse, just import the project into your workspace)

**Review by andy-goryachev-oracle — COMMENTED** [review]
I'd recommend to make further optimizations, see the code sample.

**Comment by karthikpandelu** [comment]
@andy-goryachev-oracle thanks for the code snippet and adding the Label page to the Monkey Tester.

In case of width calculation, the optimizations done in the code snippet is better since we avoid some computation. I implemented similar logic to `computePrefHeight()` method. 
With this optimization, for a case like empty text and empty graphic, the computed value of height will be 0.
This creates issue with some of the unit tests in `LabelSkinTest` because the expectation is that `computePrefHeight()` method returns single line string height for these cases. This is observed in unit test cases where we calculate preferred height and max height (max height also calls `computePrefHeight()` method).
Behaviorally there is no difference because a different method calculates the min height which returns the minimum height of a single line for the cases where text is empty and graphic is also null. But looking at the expected behavior of `computePrefHeight()` method from unit test cases, it looks like single line string height needs to be returned from this method in which case we will not be able to use this optimized `if-else` block since we have to calculate textHeight regardless of any conditions.
So if we have to optimize the `computePrefHeight()` method, the expected behavior also changes in these conditions.

Please let me know your thoughts on this.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (karthikpandelu) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java]: Updated `computePrefWidth()` method.
Please refer to my comment at the end regarding `computePrefHeight()` method optimization observations.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (karthikpandelu) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java:372]: Changed local variable name.
Please refer to my comment at the end regarding `computePrefHeight()` method optimization.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/LabeledSkinBase.java:407]: looks good, thank you.

**Review by andy-goryachev-oracle — APPROVED** [review]
looks good now, confirmed with junit and the MonkeyTester.

**Comment by karthikpandelu** [comment]
Thanks Andy.
@aghaisas and @hjohn, since there has been code changes after your approval, looks like the PR is not getting added with ready label.
Could you please review and approve the changes.

**Review by hjohn — APPROVED** [review]

**Review by aghaisas — APPROVED** [review]
Looks good!

**Comment by karthikpandelu** [comment]
@aghaisas please sponsor this PR.

**Pull request closed without merging**
